File: 2020-05-19/P17499.js
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

let length, count;
let result;

let line = 0;
rl.on('line', (input) => {
  const splitted = input.split(/\s/g).map(Number);

  if (line === 0) {
    length = splitted[0]
    count = splitted[1]
  }
  else if (line === 1) result = splitted
  else {
    const type = splitted[0];
    const num1 = splitted[1];
    const num2 = splitted[2];
    
    result = runCommand(type, result, num1, num2);

    if(line === count + 1) {
      console.log(result.join(' '))
    }
  }

  line++;
})

function runCommand(type, array, ...args) {
  switch(type) {
    case 1:
      array[args[0] - 1] += args[1];
      
      return array;
    default:
      let result = [];
      
      for (let i = 0; i < length; i++) {
        result[(type === 2 ? args[0] + i: (length - args[0]) + i) % length] = array[i]
      }

      return result;
  }
}

File: 2020-05-19/P17499.py
def runCommand(type, array, arg1, arg2 = 0):
    if type == 1:
        array[arg1 - 1] = arg2

        return array
    elif 2:
      result = [0] * length
      
      for i in range(0, length):
          result[(arg1 + i) % length] = array[i]

      return result
    else:
      result = [0] * length
      
      for i in range(0, length):
          result[(length - arg1 + i) % length] = array[i]

      return result

length, count = list(map(int, input().split()))
array = list(map(int, input().split()))

for i in range(0, count):
    arr = list(map(int, input().split()))
    
    if arr[0] == 1:
        array = runCommand(arr[0], array, arr[1], arr[2])
    else:
        array = runCommand(arr[0], array, arr[1])

print(array)